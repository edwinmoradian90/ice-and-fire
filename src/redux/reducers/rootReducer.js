import { combineReducers } from 'redux'
import mainReducer from '../reducers/mainReducer'
import categoriesReducer from '../reducers/categoriesReducer'

export default combineReducers({
    mainReducer,
    categoriesReducer
})