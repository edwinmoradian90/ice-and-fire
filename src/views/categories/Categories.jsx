import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { useDispatch, useSelector } from 'react-redux'
import {
    FETCH_CATEGORIES_REQUEST,
    FETCH_CATEGORIES_SUCCESS,
    FETCH_CATEGORIES_FAILURE,
    NEXT_PAGE,
    PREV_PAGE
} from '../../redux/constants/constants'
import { SUBMIT, RESET_SEARCH } from '../../redux/constants/searchConstants'
import Header from '../../components/header/Header'
import CategoryList from '../../components/list/CategoryList'
import Search from '../../components/search/Search'
import axios from 'axios'
import { url } from '../../config/config'
import { Container } from '../../utils/styledComponents/main'
import { Loader } from '../../components/loading/Loader'
import { Text } from '../../utils/styledComponents/main'
import { wordFormatter } from '../../utils/helpers'

function Categories(props) {
    const category = props.match.path
    const dispatch = useDispatch()
    const data = useSelector(state => state.categoriesReducer.data) || []
    const loading = useSelector(state => state.categoriesReducer.loading)
    const page = useSelector(state => state.categoriesReducer.page)
    const pageSize = useSelector(state => state.categoriesReducer.pageSize)
    const searchData = useSelector(state => state.searchReducer.searchData)
    const searchSubmitted = useSelector(state => state.searchReducer.searchSubmitted)

    const nextPage = () => {
        dispatch({
            type: NEXT_PAGE,
        })
    }

    const prevPage = () => {
        console.log('previous page')
        dispatch({
            type: PREV_PAGE,
        })
    }

    const noResults = (filteredData) => {
        if (!filteredData) {
            dispatch({
                type: NEXT_PAGE
            })
        }
    }

    useEffect(() => {
        dispatch({
            type: FETCH_CATEGORIES_REQUEST,
            payload: {
                data: ''
            }
        })

        dispatch({
            type: SUBMIT,
            payload: {
                searchSubmitted: false
            }
        })

        axios.get(`${url}/${category}?page=${wordFormatter(searchData)
            ? null
            : page
            }&pageSize=${pageSize}&name=${searchData}`)
            .then(res => {
                setTimeout(() => {
                    dispatch({
                        type: FETCH_CATEGORIES_SUCCESS,
                        payload: {
                            data: res.data,
                            currentUrl: category,
                        }
                    })
                    dispatch({ type: RESET_SEARCH })
                }, 2000)
            })
            .catch(err => {
                dispatch({
                    type: FETCH_CATEGORIES_FAILURE,
                    payload: {
                        error: err,
                        currentUrl: category
                    }
                })
            })
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [
        page,
        searchSubmitted,
        category,
        dispatch,
        pageSize
    ])

    return (
        <>
            {!loading
                ?
                (
                    <Container>
                        <Header category={category} />
                        <Search />
                        {
                            data.length !== 0
                                ?
                                <CategoryList
                                    data={data}
                                    category={category}
                                    searchData={searchData}
                                    noResults={noResults}
                                />
                                :
                                <>
                                    <Text>
                                        You know nothing, Jon Snow...
                                    </Text>
                                    <iframe
                                        title="you know nothing"
                                        src="https://giphy.com/embed/13f5iwTRuiEjjW"
                                        width="100%"
                                        height="100%"
                                        frameBorder="0"
                                        className="giphy-embed"
                                        allowFullScreen
                                    />


                                </>
                        }
                        {
                            data.length >= pageSize && data.length > 1
                                ?
                                (
                                    <div className='navigationButtons'>
                                        <Text
                                            onClick={() => prevPage()}
                                            cursor='pointer'
                                            className='prevPage'
                                            fontSize='1em'
                                            display={page > 1 ? 'flex' : 'none'}
                                        >
                                            Previous Page
                                        </Text>
                                        <Text className="pageNumber">{page}</Text>
                                        <Text
                                            cursor='pointer'
                                            onClick={() => nextPage()}
                                            className='nextPage'
                                            display={page !== pageSize ? 'flex' : 'none'}
                                        >
                                            Next Page
                                        </Text>
                                    </div>

                                )
                                :
                                null
                        }
                    </Container>
                )
                :
                (
                    <>
                        <Loader
                            loading={loading}
                        />
                    </>
                )
            }
        </>
    )
}

Categories.propTypes = {
    match: PropTypes.shape({
        path: PropTypes.string.isRequired,
    })
}

export default Categories