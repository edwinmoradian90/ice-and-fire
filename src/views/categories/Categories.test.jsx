import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import Categories from './Categories';
import rootReducer from '../../redux/reducers/rootReducer';

describe('Categories', () => {
    const store = createStore(rootReducer);
    const match = {
        path: '/categories'
    };
    it('Matches snapshot', () => {
        const component = renderer.create(
            <Provider store={store}>
                <Router>
                    <Categories match={match} />
                </Router>
            </Provider >
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});