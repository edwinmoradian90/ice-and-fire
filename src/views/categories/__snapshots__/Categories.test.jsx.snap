// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories Matches snapshot 1`] = `
<div
  className="sc-AxirZ jRTNxu"
>
  <div
    className="header"
  >
    <a
      className="sc-AxjAm irMvbk"
      href="/"
      onClick={[Function]}
    >
      <h1
        className="sc-AxhCb kYEhcE appTitle"
      >
        Ice and Fire
      </h1>
    </a>
  </div>
  <div
    className="sc-Axmtr iNEXlE search"
  >
    <input
      className="sc-AxmLO bYarhK searchInput"
      onChange={[Function]}
      placeholder="Search By Name"
      type="text"
    />
    <span
      className="sc-fzozJi dqEwWx submit"
      onClick={[Function]}
      type="submit"
    >
      <img
        alt="search"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABoAAAAaCAYAAACpSkzOAAAABmJLR0QA/wD/AP+gvaeTAAADQUlEQVRIib2WT2hcdRDHP/PeS9JdjUrUNLsmiIeiBz0FD1t3t54a2+62Gmz1IIre2pKDbUXx1PakSCUiaAULooKHpWDNZqsQadnEBqXRQpFS8NJGs1urhaT5n33v66FN2KSb5CVCv6fHvPnNZ2Z+w/CDuySrZSxlk58g2wMaBf7CdEGyMxN+Y2HT6dMz6wE5tYzTN70DoEtAnYmiUN5cu9zEjYb1QGCZigCGOxJNrfe1jpanyu1mQZdkKSAG+MCYYMzgqowz+MHX8cK5K6uCSpnks5g6CSyB8TjQOF8csCFEwmOGdbbk+39cETSSSb5g8CVwb4igK+k6cE3iooMN+p59+8ip4vACCKC8Pf1U4AQnDTatHk+/eQ6ZOdXNQmWbwceI+2s5AlcNu7zojspbt96j+smjwD5WaZmJzwLTBcO6gRWHxMy6aw7D9Y5UzPd4Xaa9QOtKQcLI0Mt3gErZ1BYT7wr5kmbNrINwA7GcxhumN7R61ZZyNplB+krwAIDZstPvA8cM/SrsI25NabSWo+BEU1/fqDe8OxFxJ720Gfsk7QyXpD6M5X96G+DvnYmhwHfTcvQasvRiP/vXn6scBXBsvOFBc3QEwkLAD+zz+e/m7wb/mIv639RXKllQl9De25DjLpZo+2HwBtwebx3GKQ89042sK1Q96JUZr67Xq5uZacsNTlX/G9mefM6MY3Kmno73DE0u1Kb29rpSLHrE0DussJKWKADKZrwfSLOOOaWWnv5TI5nkAYLgJG79RLzn7D/VB6y0I/UBpjcBNyRkqSrAGKIP4/mWSCxquZy/1MmJ9fa/JbOX1gkB8IAmjD2AV54sv6canbGRbHvUFPkZePJ/wBZJgbMlXigWF2VjipxYJ2RCAS/i6lHDHjMxJzgIKsQKxf6lzp7M6zL8S0j7geY1gC7GCwPfVxvK2dSVjTe9L+zWMl2khV4O705E3GnnVWQHw21w/ozlB9rCZnXHpekwTul8chdwyGDzMqc+RaTB3ojl+39ZF6ha13akN8uCQ4JdVL0v3IjX2Jw7Ox4GMK+aj5N5bewtnmvJD3S6OE+AHQemAPw5HloLBMJvAgBK21IP42o/4vdY70BurbC7ov8AF7ozJMqlsJ8AAAAASUVORK5CYII="
      />
    </span>
  </div>
  <div
    className="sc-AxhUy fClwJB"
  >
    You know nothing, Jon Snow...
  </div>
  <iframe
    allowFullScreen={true}
    className="giphy-embed"
    frameBorder="0"
    height="100%"
    src="https://giphy.com/embed/13f5iwTRuiEjjW"
    title="you know nothing"
    width="100%"
  />
</div>
`;
